### PR TITLE
streaming support

### LINKED ISSUE #51: Feature Request: Support streaming rather than truncate results
**problem:** all the db queries resulting from an api request include a `LIMIT` that effectively truncates the results returned when over a fixed limit.

**impact:** as the data and queries for that data grow, this may mean that customers won't always get all the data they expect, and have no way to get what they missed. this is exacerbated by the lack of guarantees around the order of data they get, meaning they might have holes in the data without realizing it that might impact the validity of how they use the data.

**proposal:** add support for pagination to the api. example reference: https://www.allphptricks.com/create-simple-pagination-using-php-and-mysqli/

### PR DESCRIPTION
**Pull request by sgratzl**
closes #51 

streams results back to the user instead of caching it locally and then send it in a bulk.

It is a bigger change since I introduced the notion of a printer which is responsible for printing in different formats. There is no change for clients. 

Following formats are supported:
 * classic = classic epidata format
 * tree = the epidata format which the first element grouped by `signal` ... no streaming
 * json = flat json array
 * csv = csv file
 * jsonl = jsonlines format (https://jsonlines.org/) which is a nice format for streaming

don't know why
```
delphi.delphi-epidata.integrations.acquisition.covidcast.test_csv_uploading.CsvUploadingTests.test_uploading: fail
```

is failing. I didn't touch those files

### CODE REVIEW

File: src/server/api.php [hunk 1]
```diff
@@ -1214,76 +1160,83 @@ function get_covid_hosp_state_timeseries($states, $dates, $issues) {
...
-    "releases" => $epidata_releases,
-    "locations" => $epidata_locations
+    "releases" => $epidata_releases->$data,
+    "locations" => $epidata_locations->$data
   );
-  return $epidata;
+  $printer->printNonStandard($epidata);
 }
 function get_meta_afhsb() {
+  // TODO
```
Comment (dfarrow0) [inline src/server/api.php:1239]: what's the thing to be done here?

File: src/server/api.php [hunk 2]
```diff
@@ -1305,124 +1258,127 @@ function get_meta_afhsb() {
...
-// connect to the database
-if(database_connect()) {
-  // select the data source
-  $source = isset($_REQUEST['source']) ? strtolower($_REQUEST['source']) : null;
+
+/*
+ * main function
+ */
+function main() {
+  $source = isset($_REQUEST['source']) ? strtolower($_REQUEST['source']) : '';
```
Comment (dfarrow0) [inline src/server/api.php:1273]: wasn't there a recent change to replace "source" with "endpoint". (not actual replacement, but making "source" and "endpoint" synonyms.) should `$_REQUEST['endpoint']` appear here too?

Comment (krivard) [reply src/server/api.php:1273]: (it's not merged yet)

File: src/server/api.php (hunk 1 above)
Comment (sgratzl) [reply src/server/api.php]: fixed

File: src/server/api.php (hunk 2 above)
Comment (sgratzl) [reply src/server/api.php]: not it has been merged into a `source-to-endpoint` branch. But I merged that one in

**Comment by dfarrow0** [comment]
looks like this other test fails too:

```
FAIL: test_caching (repos.delphi.delphi-epidata.integrations.acquisition.covidcast.test_covidcast_meta_caching.CovidcastMetaCacheTests)
Populate, query, cache, query, and verify the cache.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/src/app/repos/delphi/delphi-epidata/integrations/acquisition/covidcast/test_covidcast_meta_caching.py", line 112, in test_caching
    self.assertEqual(epidata2['result'], -2, json.dumps(epidata2))
AssertionError: 0 != -2 : {"result": 0, "message": "error: Expecting value: line 1 column 1 (char 0)"}
```

---

there's a lot going on in the pr, so i'm going to try to run the integration tests under php5 also. will update when done.

@krivard i'd love to get to php7 in prod. `delphi-epidata`, `www-epidata`, and `github-deploy-repo` are all ready. is `automation` the only thing holding this back?

**Comment by dfarrow0** [comment]
php5.4 image, streaming branch:
```
 delphi.delphi-epidata.integrations.acquisition.covid_hosp.state_timeseries.test_scenarios.AcquisitionTests.test_acquire_dataset: fail
 delphi.delphi-epidata.integrations.acquisition.covidcast.test_covidcast_meta_caching.CovidcastMetaCacheTests.test_caching: fail
 delphi.delphi-epidata.integrations.acquisition.covidcast.test_csv_uploading.CsvUploadingTests.test_uploading: fail
 delphi.delphi-epidata.integrations.client.test_delphi_epidata.DelphiEpidataPythonClientTests.test_covidcast: fail
 delphi.delphi-epidata.integrations.client.test_delphi_epidata.DelphiEpidataPythonClientTests.test_covidcast_meta: fail
 delphi.delphi-epidata.integrations.client.test_delphi_epidata.DelphiEpidataPythonClientTests.test_geo_value: fail
 delphi.delphi-epidata.integrations.server.test_api_analytics.ApiAnalyticsTests.test_analytics_update: fail
 delphi.delphi-epidata.integrations.server.test_covid_hosp.ServerTests.test_query_by_issue: fail
 delphi.delphi-epidata.integrations.server.test_covidcast.CovidcastTests.test_csv_format: fail
 delphi.delphi-epidata.integrations.server.test_covidcast.CovidcastTests.test_date_formats: error
 delphi.delphi-epidata.integrations.server.test_covidcast.CovidcastTests.test_fields: error
 delphi.delphi-epidata.integrations.server.test_covidcast.CovidcastTests.test_geo_value: error
 delphi.delphi-epidata.integrations.server.test_covidcast.CovidcastTests.test_location_timeline: error
 delphi.delphi-epidata.integrations.server.test_covidcast.CovidcastTests.test_location_wildcard: error
 delphi.delphi-epidata.integrations.server.test_covidcast.CovidcastTests.test_nullable_columns: error
 delphi.delphi-epidata.integrations.server.test_covidcast.CovidcastTests.test_raw_json_format: error
 delphi.delphi-epidata.integrations.server.test_covidcast.CovidcastTests.test_round_trip: error
 delphi.delphi-epidata.integrations.server.test_covidcast.CovidcastTests.test_temporal_partitioning: error
 delphi.delphi-epidata.integrations.server.test_covidcast_meta.CovidcastMetaTests.test_filter: error
 delphi.delphi-epidata.integrations.server.test_covidcast_meta.CovidcastMetaTests.test_round_trip: error
 delphi.delphi-epidata.integrations.server.test_covidcast_meta.CovidcastMetaTests.test_suppress_work_in_progress: error
 delphi.delphi-epidata.integrations.server.test_fluview.FluviewTests.test_round_trip: fail
 delphi.delphi-epidata.integrations.server.test_fluview_meta.FluviewMetaTests.test_round_trip: fail
```

php5.4 image, main branch from cmu-delphi:
```
✔ All 31 tests passed! 43% (189/435) coverage.
```

@sgratzl did you use any php features that aren't available in php5?

**Comment by sgratzl** [comment]
I didn't know that the server is just using php5.4. I was testing with the docker image from https://github.com/cmu-delphi/operations/blob/main/dev/docker/web/Dockerfile which is PHP 7. So, most likely with all the types and classes.

**Comment by krivard** [comment]
> get to php7 in prod

agreed, will start an asana task until we figure out what work is actually required

**Comment by dfarrow0** [comment]
yeah, not your fault for using php7 -- we were planning to go to php7 many months ago and it fizzled with everything else happening. we should discuss cost+risk+timeline for switching over to php7 (and also cost+risk of staying on php5). @krivard @korlaxxalrok 

---

@sgratzl on the php7 image, the failing test is `FAIL: test_caching (repos.delphi.delphi-epidata.integrations.acquisition.covidcast.test_covidcast_meta_caching.CovidcastMetaCacheTests)`

looks like [there's a bare return when the cache is empty](https://github.com/sgratzl/delphi-epidata/blob/06568de39cd9549978d63506a07504b2fb185b92/src/server/api.php#L969). no response is emitted, and the client is unable to parse. in this case, the desired behavior is `result` -2, `message` something about no data, and empty or absent `epidata`.

**Comment by dfarrow0** [comment]
all unit/integration tests are passing for me now. next doing some manual testing.

`curl -I 'http://127.0.0.1:10080/epidata/api.php'`
```
HTTP/1.0 500 Internal Server Error
Date: Wed, 16 Dec 2020 16:48:12 GMT
Server: Apache/2.4.38 (Debian)
X-Powered-By: PHP/7.4.13
Connection: close
Content-Type: text/html; charset=UTF-8
```

expected `200` and message about missing source:
`curl 'https://delphi.cmu.edu/epidata/api.php'`
```
{"result":-1,"message":"no data source specified"}
```

---

i tried curling the various formats with `source=fluview`, and the responses looked reasonable to me.

**Comment by sgratzl** [comment]
> all unit/integration tests are passing for me now. next doing some manual testing.

should be fixed

**Comment by sgratzl** [comment]
> I didn't know that the server is just using php5.4. I was testing with the docker image from https://github.com/cmu-delphi/operations/blob/main/dev/docker/web/Dockerfile which is PHP 7. So, most likely with all the types and classes.

So, I guess the system has been upgraded to php 7?

**Comment by krivard** [comment]
Mmmmm looks like I misinterpreted the thread to mean that had been fixed. I know Brian started looking at that yesterday, let me see if we can expedite that before this deploys to production. It's currently behind at least four indicator updates so we have some time.

**Comment by krivard** [comment]
Brian recommends we roll back. Apologies for the fire drill!

**Pull request merged**